golang_src/cmd/worker.go

```diff
@@ -0,0 +1,247 @@
+				// worker_id to nil
+				log.Printf("Updating and releasing email %s\n", email.ID)
+				mongoClient.UpdateEmail(email, &emailUpdate)
+			}
+		}
+
+		if err != nil || smtpServer == nil || email == nil {
+			if err != nil {
+				// NOTE If the datastore goes down long enough, We'd want to trigger an alert at this point,
+				//			and probably have monitoring restart the entire process via signal, as it's the safest
+				//			(non-racey) way to recover
+				log.Printf("Caught Error: %s", err.Error())
+			}
+			log.Printf("Going back to sleep for %d seconds\n", cfg.WorkerSleep)
+			time.Sleep(time.Duration(cfg.WorkerSleep) * time.Second)
+		}
+	}
+	log.Println("Finishing up...")
+}
+
+func loadConfig() (*WorkerConfig, error) {
+	viper.BindEnv("DB_HOST")
+	viper.BindEnv("DB_PORT")
+	viper.BindEnv("DB_NAME")
+	viper.BindEnv("WORKER_SLEEP")
+	viper.BindEnv("WORKER_POOL")
+	viper.BindEnv("CONNECTION_RETRIES")
+	viper.BindEnv("CONNECTION_TIMEOUT")
+	viper.BindEnv("SEND_RETRIES")
+	viper.BindEnv("SEND_RETRY_INTERVAL")
+
+	viper.SetDefault("WORKER_SLEEP", 10)
+	viper.SetDefault("WORKER_POOL", 5)
+	viper.SetDefault("CONNECTION_RETRIES", 3)
+	viper.SetDefault("CONNECTION_TIMEOUT", 30)
+	viper.SetDefault("SEND_RETRIES", 3)
+	viper.SetDefault("SEND_RETRY_INTERVAL", 600) // in seconds
+
+	missing_keys := make([]string, 0)
+	for _, key := range []string{"DB_HOST", "DB_PORT", "DB_NAME"} {
+		if !viper.IsSet(key) {
+			missing_keys = append(missing_keys, key)
+		}
+	}
+	if len(missing_keys) > 0 {
+		return nil, errors.New(fmt.Sprintf("Can't find necessary environment variable(s): %s", strings.Join(missing_keys, ", ")))
+	}
+
+	for _, key := range []string{"WORKER_SLEEP", "WORKER_POOL", "CONNECTION_RETRIES", "CONNECTION_TIMEOUT", "SEND_RETRIES", "SEND_RETRY_INTERVAL"} {
+		if viper.GetInt(key) <= 0 {
+			return nil, errors.New(fmt.Sprintf("Config %s must have a value >= 0", key))
+		}
+	}
+	return &WorkerConfig{
+		Host:              viper.GetString("db_host"),
+		Port:              viper.GetString("db_port"),
+		DatabaseName:      viper.GetString("db_name"),
+		WorkerSleep:       viper.GetInt("worker_sleep"),
+		WorkerPool:        viper.GetInt("worker_pool"),
+		ConnectionRetries: viper.GetInt("connection_retries"),
+		ConnectionTimeout: viper.GetInt("connection_timeout"),
+		SendRetries:       viper.GetInt("send_retries"),
+		SendRetryInterval: viper.GetInt("send_retry_interval"),
+	}, nil
+}
+
+func main() {
+	log.Println("Running worker")
+	workerConfig, err := loadConfig()
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	var wg sync.WaitGroup
+
+	// Workers signal on this when they quit to let the parent know they've exited
+	var quitChannels []chan bool
+	for i := 0; i < workerConfig.WorkerPool; i++ {
+		quit := make(chan bool, 1)
+		defer close(quit)
+		quitChannels = append(quitChannels, quit)
```

Comment:
No need to create a new channel for each goroutine. You could create a single channel, then close it when done and each worker would receive the close event and cleanup. (only works if you close())

Reply:
So obviously I've revealed that I'm a bit new to go by doing this 😁 Apparently that was my issue, as I tried that sans close() and everything was blocked. I'll rewire this